### ISSUE TITLE
[TW-1625] recurring tasks appear multiple times

### ISSUE THREAD

**Issue by tbabej** [tbabej/testimportfinal#1586]
_Johannes Wienke on 2015-06-18T11:51:32Z says:_ 

 Once in a while I end up with the situation that I have multiple occurrences of recurring tasks for the same day.

```
46 4d    home      nas               P7D     2 days backup NAS                                                                9.51
39 3d    home      nas               P7D     2 days backup NAS                                                                9.51
45 4d    server                      P7D     2 days update debian                                                             8.71
47 4d    home                        P7D     2 days Blumen gießen                                                             8.71
38 3d    server                      P7D     2 days update debian                                                             8.71
40 3d    home                        P7D     2 days Blumen gießen                                                             8.71
```

I could not really track down what is causing this. My initial thought that mirakel is the origin was wrong, because this also appears without mirakel.

In any case, a fix for this would really be good since recovery always takes some time. I am using recurring tasks for many things so there are often affected tasks.

**Comment by tbabej** [comment]
Migrated metadata:
```
Created: 2015-06-18T11:51:32Z
Modified: 2017-01-16T17:36:11Z```

**Comment by tbabej** [comment]
_Tomas Babej on 2015-06-19T21:01:56Z says:_ 

 Thanks for reporting!

However, I am not able to reproduce this, so I doubt this is a generic error. We will need some more information as to how to reproduce the issue. Please let us know if you were able to track down the problem.

You might be interested in reading more about reccurences, check our docs on taskwarrior.org for that.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-19T21:04:33Z says:_ 

 Tracking this down will probably be a real issue since I have no idea what is causing this. In case I end up in that situation again, I could send you my data files. Maybe this might help to find out what is going on?

**Comment by tbabej** [comment]
_Tomas Babej on 2015-06-20T13:22:31Z says:_ 

 First let's investigate the offending tasks - can you share the output of the 'info' command for the duplicate tasks? E.g.:

```
$ task 39,46 info 
```

Attributes like mask index, or creation date might tell us more.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-21T17:28:32Z says:_ 

 Unfortunately, I don't have the state available from that time anymore since I resolved the issues manually to get back to work. Once I end up in a similar situation again, I will post the required information here.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-23T19:38:47Z says:_ 

 Ok, back in the same situation. Here are the info ouputs:

```
> task 40,46 info

Name          Value
ID            40
Description   backup NAS
Status        Pending
Project       home
Recurrence    weekly
Parent task   567eb98f-db4a-46be-b182-7833c5b45052
Mask Index    2
Entered       2015-06-21 14:08:27 (2 days)
Due           2015-06-28 00:00:00
Last modified 2015-06-18 14:02:52 (5 days)
Tags          nas
Virtual tags  CHILD DUE MONTH PENDING READY TAGGED UNBLOCKED YEAR
UUID          16f321aa-4760-4230-9bcf-3c3ef037ac4b
Urgency       8.739

    project      1 *    1 =      1
    tags       0.8 *    1 =    0.8
    due      0.577 *   12 =   6.93
    age      0.005 *    2 =  0.011
                            ------
                             8.739


Name          Value
ID            46
Description   backup NAS
Status        Pending
Project       home
Recurrence    weekly
Parent task   567eb98f-db4a-46be-b182-7833c5b45052
Mask Index    2.000000
Entered       2015-06-21 07:00:01 (2 days)
Due           2015-06-28 00:00:00
Last modified 2015-06-18 14:02:52 (5 days)
Tags          nas
Virtual tags  CHILD DUE MONTH PENDING READY TAGGED UNBLOCKED YEAR
UUID          910b7b35-f9e6-4360-9271-437ad257377b
Urgency       8.739

    project      1 *    1 =      1
    tags       0.8 *    1 =    0.8
    due      0.577 *   12 =   6.93
    age      0.005 *    2 =  0.011
                            ------
                             8.739
```

**Comment by tbabej** [comment]
_Tomas Babej on 2015-06-24T13:25:05Z says:_ 

 Okay, so we can see that:

1.) both task have the same parent 567eb98f-db4a-46be-b182-7833c5b45052, which is good, since it means there are not two duplicate template tasks
2.) one task has mask index "2" and the other "2.000000". Mask index should be integer, which denotes the index of the recurrent child task (e.g. which generated task it is in order).

So the presence of 2.0000 is the culprit here I believe. Are you really sure that Mirakel is not causing this behaviour? I encountered some problems with Mirakel of this nature.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-24T13:39:21Z says:_ 

 Mirakel is not even installed anymore on my phone. It used to be connected to this account. Can this still be a relic?

**Comment by tbabej** [comment]
_Tomas Babej on 2015-06-25T16:37:02Z says:_ 

 Maybe.. the task recurs weekly.

How long since you removed the Mirakel app? Additionally, can we have a look at the parent task?

```
$ task 567eb98f-db4a-46be-b182-7833c5b45052 info
```

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-26T12:51:04Z says:_ 

 I am quite sure that I completely deleted the recurring task including all occurrences the last time I fixed this after deleting mirakel. And in any case, the instance of the recurring task with the offending mask has been created just recently without mirakel being installed anymore. Deinstallation of mirakel was several weeks ago.

```
languitar(at)milt ~> task 567eb98f-db4a-46be-b182-7833c5b45052 info

Name          Value                                      
ID            27
Description   backup NAS                                 
Status        Recurring
Project       home                                       
Recurrence    weekly
Mask          ++-                                        
Entered       2015-06-08 22:54:59 (2 weeks)
Due           2015-06-14 00:00:00                        
Last modified 2015-06-21 14:25:03 (5 days)
Tags          nas                                        
Virtual tags  MONTH OVERDUE PARENT TAGGED UNBLOCKED YEAR
UUID          567eb98f-db4a-46be-b182-7833c5b45052       
Urgency       13.89

    project      1 *    1 =      1
    tags       0.8 *    1 =    0.8
    due          1 *   12 =     12
    age      0.047 *    2 =  0.093
                            ------
                             13.89

Date                Modification                     
2015-06-18 13:47:06 Mask set to '-'.
2015-06-18 13:47:06 Mask changed from '-' to '--'.   
2015-06-18 13:47:06 Mask changed from '--' to '+-'.
2015-06-18 13:47:06 No changes made.                 
2015-06-26 14:54:38 Mask changed from '+-' to '+X'.
2015-06-26 14:54:38 Mask changed from '+X' to '+X-'. 
2015-06-26 14:54:38 Mask changed from '+X-' to '+--'.
2015-06-26 14:54:38 Mask changed from '+--' to '++-'.
2015-06-26 14:54:38 No changes made.
```

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-29T10:19:04Z says:_ 

 I just got another copy of the offending tasks. This time mask index was 3.0000. And I am quite sure that I did not see them yesterday on a different computer. So it might be related to syncing.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-29T10:43:50Z says:_ 

 Ok, another examination. All imask values in the *.data files are floating point numbers. Maybe this is really still a relic from mirakel. I tried to convert all of them to integers. However, these changes don't propagate to the task server. Therefore I tried to do a sync init, which fails with a lot of warnings about broken ancestors. So something must be completely messed up in my data.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-06-29T11:08:48Z says:_ 

 Ok, definitely a sync prblem as far as I can tell. I completely wiped everything on two clients and the task server. Afterwards I rebuilt all my tasks from scratch on one computer. All imasks are integers. I did a sync init and on the still empty computer a sync. The downloaded data contains all floating point numbers for imask values. imask values in the tx.data file on the server for this user are still integers, not float. So something must be breaking when downloading.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-07-05T15:23:08Z says:_ 

 No I am back at duplicated tasks, after a completely starting from scratch with the config files. Any way I can help to sort this out?

**Comment by tbabej** [comment]
_Paul Beckingham on 2015-07-07T15:06:30Z says:_ 

 Taskd bug.  Taskserver 1.2 should resolve this, but right now has not.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-07-07T15:15:17Z says:_ 

 Does this mean there is also no fix in git master?

**Comment by tbabej** [comment]
_Paul Beckingham on 2015-07-07T20:23:34Z says:_ 

 That's right - this is not fixed yet.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-10-26T22:55:14Z says:_ 

 Any progress on this? I had a look at the server side and implemented a custom rule to treat imasks as integers. However, in the client the mask still ends up as a floating point number. So there is probably also an issue there.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-11-23T22:18:39Z says:_ 

 This patch adds a temporary workaround on the client side to interpret imasks always as integers.

**Comment by tbabej** [comment]
_Johannes Wienke on 2015-11-30T08:55:24Z says:_ 

 It seems this also doesn't help. I still see multiple instances of the same recurring task and all my clients are patched to have integer masks.

**Comment by tbabej** [comment]
_Johannes Wienke on 2016-01-08T07:33:28Z says:_ 

 Can someone please mark this as a duplicate of TD-44. This seems to be the origin or at least these issues should be related.

**Comment by tbabej** [comment]
_Paul Beckingham on 2016-04-09T14:56:31Z says:_ 

 See TD-44 for workaround.
